src/scipp/core/dimensions.py

```diff
@@ -0,0 +1,81 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+# @author Simon Heybrock
+from .._scipp.core import Variable, DataArray, Dataset, DimensionError
+
+
+def _make_sizes(obj):
+    """
+    Makes a dictionary of dimensions labels to dimension sizes
+    """
+    return dict(zip(obj.dims, obj.shape))
+
+
+def _make_renaming_dict(dims_dict: dict = None, **names) -> dict:
+    """
+    Make a dict of renaming dimensions, using either the dict supplied or generating
+    one from the keyword arguments.
+
+    Note that this implementation fails in the unlikely case that one of the dims is
+    named `dims_dict`. An implementation that would capture this would be
+    more complicated/ugly.
+    """
+    out = {**names}
+    if dims_dict is not None:
+        if isinstance(dims_dict, dict):
+            return dims_dict
+        elif isinstance(dims_dict, str):
+            # This is the unlikely case that `dims_dict` is one of the old dimensions
+            out['dims_dict'] = dims_dict
+        else:
+            raise TypeError('A dict mapping old dimension names to new ones must be '
+                            'supplied. Got a {}.'.format(type(dims_dict)))
+    return out
+
+
+def _rename_variable(var: Variable, dims_dict: dict = None, **names) -> Variable:
+    """
+    Rename the dimensions labels of a Variable.
+    The renaming can be defined:
+       - using a dict mapping the old to new names, e.g. rename({'x': 'a', 'y': 'b'})
+       - using keyword arguments, e.g. rename(x='a', y='b')
+    """
+    return var.rename_dims(_make_renaming_dict(dims_dict, **names))
+
+
+def _rename_data_array(da: DataArray, dims_dict: dict = None, **names) -> DataArray:
+    """
+    Rename the dimensions, coordinates and attributes of a Dataset.
+    The renaming can be defined:
+       - using a dict mapping the old to new names, e.g. rename({'x': 'a', 'y': 'b'})
+       - using keyword arguments, e.g. rename(x='a', y='b')
+    """
+    renaming_dict = _make_renaming_dict(dims_dict, **names)
+    out = da.rename_dims(renaming_dict)
+    for coord in renaming_dict:
+        for meta in (out.coords, out.attrs):
+            if coord in meta:
+                if (renaming_dict[coord] in meta) and (renaming_dict[coord]
+                                                       not in renaming_dict):
```

Comment:
This is when you are trying to do something like
```Py
renamed = da.rename(dims_dict={'x': 'y', 'z': 'x'})
```
where the `x` dim is first going to be renamed to `y`, and then the `z` dim becomes the new `x`.
However, now that I am reading this again, the renaming is actually applied duing the iterations, and so I think this check is no longer needed here.

It is probably left over from a previous version where I was first scanning the whole dict of renamings, and I needed to perform some additional checks, because in the above case, it would see you are trying to rename `z` to `x` but `x` already exists. However, if you then said `x` is also being renamed because it is also found in the renaming dict, then don't raise the error. The gap in the logic is probably that this wasn't checking for the order of the renamings.

Anyway, I think I can just remove the second check and it should all still work.

Reply:
I think `rename_dims` is not supporting cycles either (though it could).